### ISSUE TITLE
Configs: Move `no-unnecessary-type-parameters` from strictTypeChecked to stylisticTypeChecked

### ISSUE THREAD

**Issue by eranhirsch** [typescript-eslint/typescript-eslint#9887]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

The rule doesn't fit the definition of `strict`:

> a superset of recommended that includes more opinionated rules **which may also catch bugs**.

Its cause and suggestions are completely erased when the code is transpiled/built, and I can't think of a case where it catches a runtime bug (a bug that would manifest in the JS version of the code). In some cases, its suggestions can lead to code that is **less** type-safe/strict, and following them can result in runtime bugs (#9885).

It seems much more suitable for the `sytlistic` config:

> additional rules that enforce consistent styling without significantly catching bugs or changing logic.

### Impacted Configurations

strictTypeChecked, stylisticTypeChecked

### Additional Info

Sorry for spamming the issues with this rule; it's the only rule causing us issues migrating to v8 in Remeda, and I don't want to blindly ignore it because I think it makes sense and is valuable. It also surfaced a few real issues in our code (https://github.com/remeda/remeda/pull/801). I appreciate these kinds of rules and want people to use as many lint rules as possible in projects, so it's vital that they are configured correctly and communicated correctly to people

**Comment by JoshuaKGoldberg** [comment]
Hey thanks for filing this - and the additional info note, it makes us happy to hear you're getting benefit from it and working to keep a lot of rules enabled!

The `no-unnecessary-type-parameters` rule _does_ catch runtime bugs, though. It's just indirect rather than direct.

The `stylistic` ruleset is around code that's truly _visually stylistic_, such as using `T[]` vs. `Array<T>`.

---

For visibility, it's common to see code like this:

```ts
function stringify<T>(text: string): T {
  return JSON.stringify(text) as T;
}

// ...

// Doesn't have a error, but should
const data: { values: number[] } = stringify(await readFile("..."));

// May crash - if only there was a type error!
console.log(data.values[0]); 
```

`no-unnecessary-type-parameters` detects the root cause that allowed the potential crash. It operates on the _logic declared in the code's types_, namely the `<T>` type parameter.

---

Anyway, thanks for filing - the rule is definitely sticking in the logical rulesets for now. You've filed a lot of great bugs on us for it 😄 so I'm hopeful once those are fixed and/or had FAQs added for, it'll be smoother.

**Comment by eranhirsch** [comment]
The real problem is the usage of `as`, and it seems there should have been a strict typescript flag to warn on the usage of `as`, and if not a typescript config, then a lint rule. It's rarely a good sign needing `as` in your code. I'd love to know if there are examples of runtime bugs this rule would prevent that don't use `as`.

**Comment by JoshuaKGoldberg** [comment]
If you don't have `no-unsafe-*` rules enabled, this is allowed too:

```ts
function stringify<T>(text: string): T {
  return JSON.stringify(text);
}
```

**Comment by JoshuaKGoldberg** [comment]
> a lint rule for casting `any` and `unknown` using `as`

Yeah 😞 I've always wanted something like that. We've looked into it a bit but haven't build one in to typescript-eslint yet: 

* #7173: accepting PRs 😁 
* #8723: we won't pick it up for typescript-eslint, but it'd be swell to see someone make a community rule!


> examples of runtime bugs this rule would prevent that don't use `as`

If you don't have `no-unsafe-*` rules enabled, this is allowed too:

```ts
function stringify<T>(text: string): T {
  return JSON.stringify(text);
}
```

Separately from runtime bugs, you can think of this rule as sibling/cousin-ish to `no-unused-vars` & the like. `no-unused-vars` doesn't catch runtime bugs, but is more than stylistic. It helps you remove unnecessary code. Much like this one!

**Comment by eranhirsch** [comment]
My concern is that outside of this conversation, there are hundreds of projects using this rule with people who are much less savvy about typescript that could run into the rule firing an error for them and needing to solve it without much context or understanding, would run into situations like those surfaced in #9885 and end up with code that is **less** strict. This is really risky for the community using this rule.

**Comment by kirkwaiblinger** [comment]
nit - the canonical example is parsing, not stringifying 🙂 
```ts
function parseJSON<T>(input: string): T {
  return JSON.parse(input);
}

const o: { foo: string } = parseJSON('unsafe inferred type parameter');

console.log(o.foo.length); // runtime error
```

(@eranhirsch note that this is already [documented on the rule page](https://typescript-eslint.io/rules/no-unnecessary-type-parameters#examples))

**Comment by kirkwaiblinger** [comment]
By contrast, the code after fixing the report in https://github.com/typescript-eslint/typescript-eslint/issues/9885 _cannot cause a runtime type error_. It is at best a false equivalence to compare the potentials for bugs between these cases.

**Comment by eranhirsch** [comment]
> I'm now wondering if there's more to be said there - that folks should, say, feel comfortable evaluating when to use ESLint inline config comments to disable rules. What do you think?

Interesting! When I work with companies, I usually turn these rules **on** *because* their developers are less proficient with TypeScript. My thinking is that they don't have access to experienced TypeScript developers to catch errors in code review, so they need more tooling to substitute for that. I use lint as an automated code reviewer that surfaces potential bugs for them.

**Comment by kirkwaiblinger** [comment]
FWIW - I have downgraded `no-unnecessary-type-parameters` to a warning in my projects rather than an error. Most (but not all!) existing code that I've run it on has been type-safe already, just with the _unnecessary_ parameter. A few have been of the `JSON.parse` intentionally unsafe type assertion variety (yuck!). None of the reports have been urgent to fix, but it's helpful to see the report in the editor in case it is easy to fix one. 

Also, to @JoshuaKGoldberg's point, IMO `eslint-disable` comments should not be stigmatized! Some rules are cut and dry, but many rules (especially in the strict-type-checked set, e.g. `no-floating-promises` or `no-unnecessary-condition`) have unavoidable false positives in edge cases (some, but not all, of which are documented on the rule page as "Limitations"), or, like this rule, just cases that don't _need_ to be fixed. An `eslint-disable` comment, with an explanation of why it's disabled, is totally appropriate.

**Comment by Josh-Cena** [comment]
> My thinking is that they don't have access to experienced TypeScript developers to catch errors in code review, so they need more tooling to substitute for that.

If they can't catch the bug, it's possible that they don't know how to fix the bug when someone points it out either.

**Comment by eranhirsch** [comment]
> If they can't catch the bug, it's possible that they don't know how to fix the bug when someone points it out either.

But it gives them an opportunity to read about it, ask about it, and learn about it. It's a speedbump.

**Comment by eranhirsch** [comment]
To me, lint (and type checking, which could be considered a linter too) is a form of testing. The more rules you have enabled, the higher the likelihood that code that is "all green" is safe(r) to ship, and if the code isn't "all green" then they have some place to go look for an answer. That's why lint rules that are often disabled are extra dangerous, as they make  engineers complacent.

**Comment by kirkwaiblinger** [comment]
@eranhirsch, I am not sure that this conversation is going anywhere useful anymore.

You've proposed removing a rule from the strict config, which we've declined based on our judgement about what the explicitly stated intention behind the strict-type-checked config is.

You've said that your usage of that rule is different than the stated intention, so we've suggested either not using it, using judicious disable comments as appropriate, or disabling/downgrading the specific rule that you have issues with. But you've indicated you're averse to each of those...

We cannot fix every rule report to be 100% accurate, sometimes due to technical limitations (which are sometimes even documented in Limitations sections on the rule pages), and sometimes due to judgement calls that have to be made, between over-reporting and under-reporting. Again, you'll find many more of these sort of cases in the strict presets than the recommended.

We genuinely strive to be helpful, but, I'm just not sure what we can do to help you here based on this conversation. Is there something concrete you can suggest?

**Comment by eranhirsch** [comment]
I'm sorry. My intention was not to criticize the decision or try to get it overturned. The discussion went into more theoretical points, and I felt that it was an OK platform to discuss them.

**Comment by kirkwaiblinger** [comment]
All good! And I don't mean to be dismissive, apologies to come off that way. Just coming at gh issues from the lens of trying to find action items to be able to help you and/or future users in your shoes! We do tend to be more chatty in the discord, too 🙂
